nifi-nar-bundles/nifi-irc-bundle/nifi-irc-processors/src/main/java/org/apache/nifi/processors/irc/handlers/ConsumerEventHandler.java

```diff
@@ -0,0 +1,127 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.processors.irc.handlers;
+
+import net.engio.mbassy.listener.Handler;
+import net.engio.mbassy.listener.Invoke;
+import org.apache.nifi.flowfile.FlowFile;
+import org.apache.nifi.logging.ComponentLog;
+import org.apache.nifi.processor.ProcessContext;
+import org.apache.nifi.processor.ProcessSession;
+import org.apache.nifi.processor.ProcessSessionFactory;
+import org.apache.nifi.processor.exception.FlowFileAccessException;
+import org.apache.nifi.processor.io.OutputStreamCallback;
+import org.apache.nifi.processors.irc.ConsumeIRC;
+import org.apache.nifi.util.StopWatch;
+import org.kitteh.irc.client.library.event.channel.ChannelMessageEvent;
+import org.kitteh.irc.client.library.event.helper.MessageEvent;
+import org.kitteh.irc.client.library.event.user.PrivateMessageEvent;
+import org.kitteh.irc.client.library.util.Format;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+
+public class ConsumerEventHandler extends EventHandler {
+
+    public ConsumerEventHandler(ProcessContext context, ProcessSessionFactory sessionFactory, ComponentLog logger) {
+        super(context, sessionFactory, logger);
+    }
+
+    @Handler(delivery = Invoke.Asynchronously)
+    protected void onPrivateMessageReceived(PrivateMessageEvent event) {
+        logger.info("Received private message '{}' from {} while waiting for messages on {} ",
+                new Object[] {event.getMessage(), event.getActor().getName(), context.getProperty(ConsumeIRC.IRC_CHANNEL).getValue()});
+        if (context.getProperty(ConsumeIRC.IRC_PROCESS_PRIV_MESSAGES).asBoolean()) {
+            turnEventIntoFlowFile(event);
+        } else {
+            event.sendReply(String.format("Hi %s. Thank you for your message but I am not looking to chat with strangers.",
+                    String.valueOf(event.getActor().getNick())));
+        }
+    }
+
+    @Handler(delivery = Invoke.Asynchronously)
+    protected void onChannelMessageReceived(ChannelMessageEvent event) {
+        // verify if the message was sent to the channel the processor is consuming
+        if (event.getChannel().getName().equals(context.getProperty(ConsumeIRC.IRC_CHANNEL).getValue())) {
+
+            logger.info("Received message '{}'  on channel {} while waiting for messages on {} ",
+                    new Object[]{event.getMessage(), event.getChannel().getName(), context.getProperty(ConsumeIRC.IRC_CHANNEL).getValue()});
+            turnEventIntoFlowFile(event);
+        }
+    }
+
+    private void turnEventIntoFlowFile(final MessageEvent messageEvent) {
```

Comment:
I'm not super familiar with this irc library, but I think that what is going on here is a bit "unorthodox". I think that this irc client allows you to register callbacks, and when a message arrives, it calls your callback. If I'm reading this right this class has stashed away the session that was passed into the first onTrigger(), and then used to create new flowfiles OUTSIDE the onTrigger method of the processor, which is certainly not the pattern for most of the other "GetX" processors I've seen (I've seen: create a queue between the service and the processor, and when onTrigger() is called consumes from that queue). @joewitt  - do you have any cautionary tales about using ProcessSession in this way?